File: windows/src/RealtimeTranslator.Core/Realtime/RealtimeSubtitleProcessor.cs
using System;
using System.Collections.Generic;
using RealtimeTranslator.Core.Audio;
using RealtimeTranslator.Core.OpenAI;

namespace RealtimeTranslator.Core.Realtime;

internal abstract record RealtimeSubtitleRoutingAction
{
    internal sealed record None : RealtimeSubtitleRoutingAction;

    internal sealed record Select(RealtimeTranslationOutputLanguage? Target)
        : RealtimeSubtitleRoutingAction;

    internal sealed record Switch(RealtimeTranslationOutputLanguage? Target)
        : RealtimeSubtitleRoutingAction;
}

internal sealed record RealtimeSubtitleProcessingResult(
    IReadOnlyList<RealtimeSubtitleUpdate> Updates,
    RealtimeSubtitleRoutingAction RoutingAction,
    RealtimeSubtitleUpdate IngestedUpdate,
    bool IsSourceUpdate);

internal sealed class RealtimeSubtitleProcessor
{
    private readonly RealtimeSubtitleAssembler _assembler = new();
    private readonly SourceBoundaryTracker _sourceBoundaryTracker = new();

    internal LanguagePair? ActiveLanguagePair { get; private set; }

    internal string RoutingSourceText { get; private set; } = string.Empty;

    private RealtimeTranslationOutputLanguage? _selectedTranslationTarget;
    private int _reverseEvidenceCount;
    private readonly HashSet<string> _handledFailedSourceKeys = new(StringComparer.Ordinal);

    internal int CurrentSourceLength => _assembler.CurrentSourceLength;

    internal int SegmentGeneration => _assembler.SegmentGeneration;

    internal void BeginEpoch(int epoch, LanguagePair pair)
    {
        _assembler.SetLanguagePair(pair);
        _assembler.BeginNewEpoch(epoch);
        ClearBoundaryCandidate();
        RoutingSourceText = string.Empty;
        ActiveLanguagePair = pair;
        _selectedTranslationTarget = null;
        _reverseEvidenceCount = 0;
        _handledFailedSourceKeys.Clear();
    }

    internal void DeactivateLanguagePair()
    {
        ActiveLanguagePair = null;
    }

    internal void ClearBoundaryCandidate()
    {
        _sourceBoundaryTracker.Reset();
        _assembler.SetBoundaryCandidatePending(false);
    }

    internal void ResetRoutingForNextSegment()
    {
        RoutingSourceText = string.Empty;
        _selectedTranslationTarget = null;
        _reverseEvidenceCount = 0;
        ClearBoundaryCandidate();
        _assembler.ExpectLane(null);
    }

    internal RealtimeSubtitleUpdate DiscardUnconfirmed()
    {
        _assembler.DiscardUnconfirmed();
        ClearBoundaryCandidate();
        RoutingSourceText = string.Empty;
        _selectedTranslationTarget = null;
        _reverseEvidenceCount = 0;
        return new RealtimeSubtitleUpdate(
            string.Empty,
            string.Empty,
            IsTranslationCurrent: false,
            ShouldFinalize: false,
            _assembler.SegmentGeneration,
            IsInvalidation: true);
    }

    internal RealtimeSubtitleUpdate? DiscardFailedSource(string? itemId, string? eventId)
    {
        var key = itemId ?? eventId;
        if (key is not null && _handledFailedSourceKeys.Contains(key))
        {
            return null;
        }

        if (!_assembler.HasUnconfirmedContent)
        {
            return null;
        }

        if (key is not null)
        {
            _handledFailedSourceKeys.Add(key);
        }

        return DiscardUnconfirmed();
    }

    internal RealtimeSubtitleUpdate? Tick(DateTimeOffset now) => _assembler.Tick(now);

    internal RealtimeSubtitleProcessingResult? Process(
        RealtimeTranslationStreamEvent streamEvent,
        DateTimeOffset now)
    {
        ArgumentNullException.ThrowIfNull(streamEvent);

        if (streamEvent.Event is RealtimeTranslationServerEvent.InputTranscriptFailed)
        {
            return null;
        }

        var deltaStart = _assembler.CurrentSourceLength;
        var sourceDelta = streamEvent.Event
            is RealtimeTranslationServerEvent.InputTranscriptDelta source
            && streamEvent.Lane.IsSource
            ? source.Delta
            : null;
        var update = _assembler.Ingest(streamEvent, now);
        if (update is null)
        {
            return null;
        }

        var result = new RealtimeSubtitleProcessingResult(
            [update.Value],
            new RealtimeSubtitleRoutingAction.None(),
            update.Value,
            sourceDelta is not null);
        if (sourceDelta is null)
        {
            return result;
        }

        return EvaluateSourceRouting(sourceDelta, deltaStart, now, result);
    }

    /// <summary>原文 delta の文字種から発話言語を決め、逆側 target へ音声を切り替える。</summary>
    private RealtimeSubtitleProcessingResult EvaluateSourceRouting(
        string delta,
        int deltaStart,
        DateTimeOffset now,
        RealtimeSubtitleProcessingResult result)
    {
        if (ActiveLanguagePair is not { } pair)
        {
            return result;
        }

        RoutingSourceText = RoutingSourceTextWindow.Trim(RoutingSourceText + delta, pair);
        var evidence = SpokenLanguageDetector.RecentEvidence(RoutingSourceText, pair);
        var selection = TranslationTargetSelector.Select(
            pair,
            _selectedTranslationTarget,
            _reverseEvidenceCount,
            evidence);
        _reverseEvidenceCount = selection.ReverseEvidenceCount;

        if (_selectedTranslationTarget is not { } currentTarget)
        {
            if (selection.Target is not { } target)
            {
                return result;
            }

            _selectedTranslationTarget = target;
            _sourceBoundaryTracker.Reset();
            _assembler.SetBoundaryCandidatePending(false);
            _assembler.ExpectLane(target);
            return result with { RoutingAction = new RealtimeSubtitleRoutingAction.Select(target) };
        }

        if (selection.Target == currentTarget)
        {
            if (pair.Counterpart(currentTarget) is { } currentLanguage)
            {
                _sourceBoundaryTracker.Observe(
                    _assembler.CurrentSourceText,
                    deltaStart,
                    _assembler.SegmentGeneration,
                    pair,
                    currentLanguage,
                    selection.ReverseEvidenceCount);
                _assembler.SetBoundaryCandidatePending(
                    _sourceBoundaryTracker.CandidateOffset is not null);
            }

            return result;
        }

        if (pair != LanguagePair.EnEs
            && pair.Counterpart(currentTarget) is { } currentLanguageForSwitch)
        {
            _sourceBoundaryTracker.Observe(
                _assembler.CurrentSourceText,
                deltaStart,
                _assembler.SegmentGeneration,
                pair,
                currentLanguageForSwitch,
                0);
        }

        var offset = _sourceBoundaryTracker.CandidateOffset ?? deltaStart;
        var split = _assembler.SplitForLanguageSwitch(offset, now);
        _sourceBoundaryTracker.Reset();
        RoutingSourceText = RoutingSourceTextWindow.Trim(
            _assembler.CurrentSourceText,
            pair);
        _selectedTranslationTarget = selection.Target;
        _reverseEvidenceCount = 0;
        _assembler.ExpectLane(selection.Target);
        return result with
        {
            Updates = split.Finalized is { } finalized
                ? [finalized, split.Current]
                : [split.Current],
            RoutingAction = new RealtimeSubtitleRoutingAction.Switch(selection.Target),
        };
    }
}

File: RealtimeTranslator/Realtime/RealtimeSubtitleProcessor.swift
import Foundation

enum RealtimeSubtitleRoutingAction: Equatable, Sendable {
    case none
    case select(RealtimeTranslationOutputLanguage?)
    case `switch`(RealtimeTranslationOutputLanguage?)
}

struct RealtimeSubtitleProcessingResult: Equatable, Sendable {
    var updates: [RealtimeSubtitleUpdate]
    var routingAction: RealtimeSubtitleRoutingAction
    var ingestedUpdate: RealtimeSubtitleUpdate
    var isSourceUpdate: Bool
}

struct RealtimeSubtitleProcessor: Sendable {
    private var assembler = RealtimeSubtitleAssembler()
    private(set) var activeLanguagePair: LanguagePair?
    private(set) var routingSourceText = ""
    private var selectedTranslationTarget: RealtimeTranslationOutputLanguage?
    private var reverseEvidenceCount = 0
    private var sourceBoundaryTracker = SourceBoundaryTracker()
    private var handledFailedSourceKeys = Set<String>()

    var currentSourceLength: Int { assembler.currentSourceLength }
    var currentSegmentGeneration: Int { assembler.currentSegmentGeneration }
    var isCurrentSegmentTainted: Bool { assembler.isCurrentSegmentTainted }
    var hasSelectedTranslationTarget: Bool { selectedTranslationTarget != nil }

    mutating func beginEpoch(_ epoch: Int, pair: LanguagePair) {
        assembler.beginNewEpoch(epoch)
        clearBoundaryCandidate()
        routingSourceText = ""
        activeLanguagePair = pair
        selectedTranslationTarget = nil
        reverseEvidenceCount = 0
        handledFailedSourceKeys.removeAll()
        assembler.setLanguagePair(pair)
    }

    mutating func deactivateLanguagePair() {
        activeLanguagePair = nil
    }

    mutating func clearBoundaryCandidate() {
        sourceBoundaryTracker.reset()
        assembler.setBoundaryCandidatePending(false)
    }

    mutating func resetRoutingForNextSegment() {
        routingSourceText = ""
        selectedTranslationTarget = nil
        reverseEvidenceCount = 0
        clearBoundaryCandidate()
        assembler.expectLane(nil)
    }

    mutating func discardUnconfirmed() -> RealtimeSubtitleUpdate {
        assembler.discardUnconfirmed()
        clearBoundaryCandidate()
        routingSourceText = ""
        selectedTranslationTarget = nil
        reverseEvidenceCount = 0
        return RealtimeSubtitleUpdate(
            sourceText: "",
            translatedText: "",
            isTranslationCurrent: false,
            shouldFinalize: false,
            segmentGeneration: assembler.currentSegmentGeneration,
            isInvalidation: true
        )
    }

    mutating func discardFailedSource(itemID: String?, eventID: String?) -> RealtimeSubtitleUpdate? {
        let key = itemID ?? eventID
        if let key, handledFailedSourceKeys.contains(key) {
            return nil
        }
        guard assembler.hasUnconfirmedContent else { return nil }
        if let key {
            handledFailedSourceKeys.insert(key)
        }
        return discardUnconfirmed()
    }

    mutating func markAudioLoss(now: Date) -> RealtimeSubtitleUpdate {
        assembler.markAudioLoss(now: now)
        clearBoundaryCandidate()
        routingSourceText = ""
        selectedTranslationTarget = nil
        reverseEvidenceCount = 0
        assembler.expectLane(nil)
        return RealtimeSubtitleUpdate(
            sourceText: "",
            translatedText: "",
            isTranslationCurrent: false,
            shouldFinalize: false,
            segmentGeneration: assembler.currentSegmentGeneration,
            isInvalidation: true
        )
    }

    mutating func tick(now: Date) -> RealtimeSubtitleUpdate? {
        assembler.tick(now: now)
    }

    mutating func process(
        _ streamEvent: RealtimeTranslationStreamEvent,
        now: Date,
        isReplay: Bool = false
    ) -> RealtimeSubtitleProcessingResult? {
        if case .inputTranscriptFailed = streamEvent.event {
            return nil
        }
        let deltaStart = assembler.currentSourceLength
        let sourceDelta: String?
        if case .inputTranscriptDelta(let delta, _, _) = streamEvent.event,
           streamEvent.lane.isSource {
            sourceDelta = delta
        } else {
            sourceDelta = nil
        }
        guard let update = assembler.ingest(streamEvent, now: now, isReplay: isReplay) else {
            return nil
        }
        let result = RealtimeSubtitleProcessingResult(
            updates: [update],
            routingAction: .none,
            ingestedUpdate: update,
            isSourceUpdate: sourceDelta != nil
        )
        guard let sourceDelta else { return result }
        return evaluateSourceRouting(delta: sourceDelta, deltaStart: deltaStart, now: now, result: result)
    }

    private mutating func evaluateSourceRouting(
        delta: String,
        deltaStart: Int,
        now: Date,
        result: RealtimeSubtitleProcessingResult
    ) -> RealtimeSubtitleProcessingResult {
        var result = result
        guard let pair = activeLanguagePair else { return result }

        routingSourceText = RoutingSourceTextWindow.trim(
            routingSourceText + delta,
            pair: pair
        )
        let evidence = SpokenLanguageDetector.recentEvidence(
            in: routingSourceText,
            pair: pair
        )
        var oppositeRun: OppositeScriptRun?
        if pair != .enEs,
           let currentTarget = selectedTranslationTarget,
           let currentLanguage = pair.counterpart(of: currentTarget)
        {
            sourceBoundaryTracker.observe(
                segmentSource: assembler.currentSourceText,
                deltaStart: deltaStart,
                segmentGeneration: assembler.currentSegmentGeneration,
                pair: pair,
                currentLanguage: currentLanguage,
                reverseEvidenceCount: 0
            )
            oppositeRun = sourceBoundaryTracker.oppositeScriptRun(
                in: assembler.currentSourceText
            )
        }
        let selection = TranslationTargetSelector.select(
            pair: pair,
            currentTarget: selectedTranslationTarget,
            reverseEvidenceCount: reverseEvidenceCount,
            evidence: evidence,
            oppositeRun: oppositeRun
        )
        reverseEvidenceCount = selection.reverseEvidenceCount

        guard selectedTranslationTarget != nil else {
            guard let target = selection.target else { return result }
            selectedTranslationTarget = target
            sourceBoundaryTracker.reset()
            assembler.setBoundaryCandidatePending(false)
            assembler.expectLane(target)
            result.routingAction = .select(target)
            return result
        }

        guard let currentTarget = selectedTranslationTarget else {
            return result
        }
        guard let target = selection.target, target != currentTarget else {
            if pair == .enEs, let currentLanguage = pair.counterpart(of: currentTarget) {
                sourceBoundaryTracker.observe(
                    segmentSource: assembler.currentSourceText,
                    deltaStart: deltaStart,
                    segmentGeneration: assembler.currentSegmentGeneration,
                    pair: pair,
                    currentLanguage: currentLanguage,
                    reverseEvidenceCount: selection.reverseEvidenceCount
                )
            }
            assembler.setBoundaryCandidatePending(
                sourceBoundaryTracker.candidateOffset != nil
            )
            return result
        }

        let offset = sourceBoundaryTracker.candidateOffset ?? deltaStart
        let split = assembler.splitForLanguageSwitch(at: offset, now: now)
        sourceBoundaryTracker.reset()
        routingSourceText = RoutingSourceTextWindow.trim(
            assembler.currentSourceText,
            pair: pair
        )
        selectedTranslationTarget = target
        reverseEvidenceCount = 0
        assembler.expectLane(target)
        result.updates = split.finalized.map { [$0, split.current] } ?? [split.current]
        result.routingAction = .switch(target)
        return result
    }
}
